### PR TITLE
Feat: planStage

### PR DESCRIPTION
**Pull request by wildeslam**
- add field stage to ReplayPlan
- update local plan's stageInfo in different stage 
- Report plan's stages regularly by ScheduledThreadPoolExecutor

```mermaid
graph LR
subgraph Init
BuildPlan-->SavePlan
SavePlan-->InitReport
end
InitReport-->LoadConfig
subgraph PreLoad
LoadConfig-->LoadCase
LoadCase-->BuildContext
end
BuildContext-->Run
Run-->Finish
```

### CODE REVIEW

File: arex-schedule-web-api/src/main/java/com/arextest/schedule/planexecution/PlanStageMonitorImpl.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
+/**
+ * @author wildeslam.
+ * @create 2023/7/25 20:03
+ */
+@Slf4j
+@Component
+public class PlanStageMonitorImpl implements PlanExecutionMonitor {
```
Comment (QizhengMo) [inline arex-schedule-web-api/src/main/java/com/arextest/schedule/planexecution/PlanStageMonitorImpl.java:25]: Can this be done in the existing execution monitor?
adding another internal class should be enough

Comment (wildeslam) [reply arex-schedule-web-api/src/main/java/com/arextest/schedule/planexecution/PlanStageMonitorImpl.java:25]: I thought so too at first. But the calling timing of register&deregister function for two monitors are different.

File: arex-schedule-web-api/src/main/java/com/arextest/schedule/dao/mongodb/ReplayPlanRepository.java [hunk 2]
```diff
@@ -65,4 +67,13 @@ public List<ReplayPlan> timeoutPlanList(Duration offsetDuration, Duration maxDur
         List<ReplayPlanCollection> replayPlanCollections = mongoTemplate.find(query, ReplayPlanCollection.class);
         return replayPlanCollections.stream().map(ReplayPlanConverter.INSTANCE::dtoFromDao).collect(Collectors.toList());
     }
+
+    public boolean updateStage(ReplayPlan replayPlan) {
+        Query query = Query.query(Criteria.where(DASH_ID).is(replayPlan.getId()));
+        Update update = MongoHelper.getUpdate();
+        update.set(REPLAY_PLAN_STAGE_LIST, replayPlan.getReplayPlanStageList());
+        mongoTemplate.findAndModify(query, update, FindAndModifyOptions.options().returnNew(false),
```
Comment (coryhh) [inline arex-schedule-web-api/src/main/java/com/arextest/schedule/dao/mongodb/ReplayPlanRepository.java:75]: can use default FindAndModifyOptions

File: arex-schedule-web-api/src/main/java/com/arextest/schedule/model/dao/mongodb/StageInfoBaseCollection.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+import lombok.Data;
+import org.springframework.data.mongodb.core.mapping.Document;
+
+/**
+ * @author wildeslam.
+ * @create 2023/7/24 16:38
+ */
+@Data
+@Document
+public class StageInfoBaseCollection {
```
Comment (coryhh) [inline arex-schedule-web-api/src/main/java/com/arextest/schedule/model/dao/mongodb/StageInfoBaseCollection.java:12]: need extends ModelBase?

File: arex-schedule-web-api/src/main/java/com/arextest/schedule/model/ReplayPlan.java [hunk 4]
```diff
@@ -109,4 +110,6 @@ public void enqueueBizLog(BizLog log) {
     public void buildActionItemMap() {
         this.getReplayActionItemList().forEach(replayActionItem -> this.actionItemMap.put(replayActionItem.getId(), replayActionItem));
     }
+
+    private List<ReplayPlanStageInfo> replayPlanStageList;
 }
```
Comment (coryhh) [inline arex-schedule-web-api/src/main/java/com/arextest/schedule/model/ReplayPlan.java:115]: code style, field before method

File: arex-schedule-web-api/src/main/java/com/arextest/schedule/service/PlanConsumePrepareService.java [hunk 5]
```diff
@@ -59,6 +59,7 @@ private int saveAllActionCase(List<ReplayActionItem> replayActionItemList) {
         long end;
 
         for (ReplayActionItem replayActionItem : replayActionItemList) {
+            // TODO: 2023/7/20 rerun
```
Comment (coryhh) [inline arex-schedule-web-api/src/main/java/com/arextest/schedule/service/PlanConsumePrepareService.java:62]: can remove?

File: arex-schedule-web-api/src/main/java/com/arextest/schedule/beans/ExecutorServiceConfiguration.java [hunk 6]
```diff
@@ -92,4 +92,14 @@ public ScheduledExecutorService monitorScheduler() {
 
         return new ScheduledThreadPoolExecutor(1, threadFactory);
     }
+
+    @Bean
+    public ScheduledExecutorService updatePlanStageScheduler() {
```
Comment (QizhengMo) [inline arex-schedule-web-api/src/main/java/com/arextest/schedule/beans/ExecutorServiceConfiguration.java:97]: remove unused pool

File: arex-schedule-web-api/src/main/java/com/arextest/schedule/planexecution/PlanExecutionMonitorImpl.java [hunk 7]
```diff
@@ -40,9 +45,11 @@ public class PlanExecutionMonitorImpl implements PlanExecutionMonitor {
     @Value("${arex.schedule.bizLog.secondToSave}")
     private long LOG_TIME_GAP_TO_SAVE_CHECK_BY_SEC;
 
-    private final RedisCancelMonitor cancelMonitor = new RedisCancelMonitor();
     private final ProgressMonitor progressMonitor = new ProgressMonitor();
     private final BizLoggerMonitor bizLoggerMonitor = new BizLoggerMonitor();
+    private final RedisCancelMonitor redisCancelMonitor = new RedisCancelMonitor();
+    private static final ConcurrentHashMap<String, List<ReplayPlanStageInfo>> replayPlanStageListMap =
```
Comment (QizhengMo) [inline arex-schedule-web-api/src/main/java/com/arextest/schedule/planexecution/PlanExecutionMonitorImpl.java:51]: Can be done by Map<String, int>, we only need the list size to detect modification

File: arex-schedule-web-api/src/main/java/com/arextest/schedule/model/dao/mongodb/StageInfoBaseCollection.java (hunk 3 above)
Comment (wildeslam) [reply arex-schedule-web-api/src/main/java/com/arextest/schedule/model/dao/mongodb/StageInfoBaseCollection.java:12]: unused class

File: arex-schedule-web-api/src/main/java/com/arextest/schedule/model/ReplayPlan.java (hunk 4 above)
Comment (wildeslam) [reply arex-schedule-web-api/src/main/java/com/arextest/schedule/model/ReplayPlan.java:115]: Resolving conflict causes this problem.

File: arex-schedule-web-api/src/main/java/com/arextest/schedule/planexecution/PlanExecutionMonitorImpl.java [hunk 8]
```diff
@@ -152,4 +159,32 @@ private void flushLogs(ReplayPlan replayPlan) {
...
+        for (; index < replayPlan.getReplayPlanStageList().size(); index++) {
+            if (replayPlan.getReplayPlanStageList().get(index).getStageStatus() == StageStatusEnum.PENDING.getCode()) {
+                break;
+            }
+        }
+        ReplayPlanStageInfo cancelStage = new ReplayPlanStageInfo();
+        cancelStage.setStageStatus(StageStatusEnum.SUCCEEDED.getCode());
+        cancelStage.setStageType(PlanStageEnum.CANCEL.getCode());
+        cancelStage.setStageName(PlanStageEnum.CANCEL.name());
+        replayPlan.getReplayPlanStageList().add(index - 1, cancelStage);
```
Comment (lucas-myx) [inline arex-schedule-web-api/src/main/java/com/arextest/schedule/planexecution/PlanExecutionMonitorImpl.java:187]: `index`如果未执行for循环的话，这里就是 `add(-1, cancelStage)` , 会报 `indexOutOfBound`的异常，
建议最好判断下

File: arex-schedule-web-api/src/main/java/com/arextest/schedule/progress/impl/UpdateResultProgressEventImpl.java [hunk 9]
```diff
@@ -76,6 +94,63 @@ public void onReplayPlanTerminate(String replayId) {
...
+            boolean lastSubStageSucceeded = stageStatus == StageStatusEnum.SUCCEEDED &&
+                parentStage.getSubStageList().get(parentStage.getSubStageList().size() - 1) == stageType.getCode();
+            // when any subStage fails, fail its parent stage
+            if (firstSubStageOnGoing || lastSubStageSucceeded || stageStatus == StageStatusEnum.FAILED) {
+                updateStage(stageInfoList, parentStage, stageStatus, startTime, endTime, null);
+            }
+        } else {
+            for (StageBaseInfo stage : stageInfoList) {
+                if (stageType.getCode() == stage.getStageType()) {
+                    stageBaseInfo = stage;
```
Comment (lucas-myx) [inline arex-schedule-web-api/src/main/java/com/arextest/schedule/progress/impl/UpdateResultProgressEventImpl.java:132]: `break;` ?

File: arex-schedule-web-api/src/main/java/com/arextest/schedule/planexecution/PlanExecutionMonitorImpl.java (hunk 8 above)
Comment (wildeslam) [reply arex-schedule-web-api/src/main/java/com/arextest/schedule/planexecution/PlanExecutionMonitorImpl.java]: 已修复

**Review by mr3 — APPROVED** [review]

**Review by QizhengMo — APPROVED** [review]

**Pull request merged**